### ISSUE TITLE
Is there a host platform macro for cxx_genrule?

### ISSUE THREAD

**Issue by njlr** [facebook/buck#2039]
During cross-compilation, we have cases where a tool is being built that will run on the _host_ machine, but the library is being built for another _target_ machine. 

For example, you might be compiling a project on macOS to run on Linux that uses protobuf. We need to compile the libraries for Linux (target) but the `protoc` tool needs to be compiled for macOS (host) so that it can be run during the build. 

Is there a macro available in `cxx_genrule` to allow me to get the host platform?

**Comment by carljparker** [comment]
Buck doesn't provide such a macro. Philosophically, Buck is intended to be cross-platform by design and so is agnostic about the platform on which it is running. 

That said, I understand how your use case requires that you be able to query the host platform. Would it work to check the string printed by the `uname` utility? It prints a string that corresponds to the operating system--and it _does_ distinguish between, say, FreeBSD and macOS; on my MacBook Pro, it prints `Darwin`.

**Comment by njlr** [comment]
Yes, a combination of `cxx_genrule` and `uname` could probably do the job. Ideally though, this would be done by Buck automatically. 

So for example, with this Buck file: 

```python=
cxx_binary(
  name = 'tool',
  srcs = [
    'tool.cpp', 
  ], 
)

genrule(
  name = 'foo-cpp', 
  out = 'foo.cpp', 
  cmd = '$(exe :tool) > $OUT', 
)

cxx_library(
  name = 'foo',
  exported_headers = {
    'foo.hpp', 'foo.hpp', 
  }, 
  srcs = [
    ':foo-cpp', 
  ],  
)
```

If I am on host Linux and target Darwin by calling `buck build //:foo#macosx-x86_64`, I would expect the following artefacts: 

 - `//:tool` compiled for Linux (because it is run locally to generate `//:foo-cpp`)
 - `//:foo-cpp` generated by calling `//:tool` (because it is required by `//:foo`)
 - `//:foo` compiled for Darwin (because that's what I requested)

**Comment by cjhopman** [comment]
What you wrote under "I expect" is exactly what happens, though, isn't it? An exe expanded in a genrule isn't going to propagate the cxx platform from some random cxx_library that depends on it, it's going to use the builds default platform.
